prov/lnx/src/lnx_ep.c

```diff
@@ -565,69 +474,43 @@ struct fi_ops_cm lnx_cm_ops = {
 	.shutdown = fi_no_shutdown,
 };
 
-static int lnx_open_eps(struct local_prov *prov, struct fi_info *info,
-			void *context, size_t fclass, struct lnx_ep *lep)
+static int
+lnx_open_core_eps(struct lnx_ep *lep, void *context, size_t fclass)
 {
-	int rc = 0;
-	struct local_prov_ep *ep;
-	struct dlist_entry *tmp;
-	struct ofi_bufpool_attr bp_attrs = {};
-	struct lnx_srx_context *ctxt;
+	int i, rc = 0;
+	struct lnx_core_domain *cd;
+	struct lnx_core_ep *cep;
 
-	ctxt = calloc(1, sizeof(*ctxt));
-	if (!ctxt)
-		return -FI_ENOMEM;
+	/* Since we kept the shm domain at the head of the array, the shm
+	 * endpoints will be at the head of the endpoint array
+	 */
+	for (i = 0; i < lep->le_domain->ld_num_doms; i++) {
+		cd = &lep->le_domain->ld_core_domains[i];
 
-	dlist_foreach_container_safe(&prov->lpv_prov_eps,
-		struct local_prov_ep, ep, entry, tmp) {
-		if (fclass == FI_CLASS_EP) {
-			rc = fi_endpoint(ep->lpe_domain, ep->lpe_fi_info,
-					 &ep->lpe_ep, context);
-		} else {
-			/* update endpoint attributes with whatever is being
-			 * passed from the application
-			 */
-			if (ep->lpe_fi_info && info) {
-				ep->lpe_fi_info->ep_attr->tx_ctx_cnt =
-					info->ep_attr->tx_ctx_cnt;
-				ep->lpe_fi_info->ep_attr->rx_ctx_cnt =
-					info->ep_attr->rx_ctx_cnt;
-			}
-
-			ep->lpe_txc = calloc(info->ep_attr->tx_ctx_cnt,
-					sizeof(*ep->lpe_txc));
-			ep->lpe_rxc = calloc(info->ep_attr->rx_ctx_cnt,
-					sizeof(*ep->lpe_rxc));
-			if (!ep->lpe_txc || !ep->lpe_rxc)
-				return -FI_ENOMEM;
-
-			rc = fi_scalable_ep(ep->lpe_domain, ep->lpe_fi_info,
-					    &ep->lpe_ep, context);
-		}
-		if (rc)
-			return rc;
+		cep = &lep->le_core_eps[i];
 
-		ctxt->srx_lep = lep;
-		ctxt->srx_cep = ep;
-
-		ep->lpe_srx.ep_fid.fid.context = ctxt;
-		ep->lpe_srx.ep_fid.fid.fclass = FI_CLASS_SRX_CTX;
-		ofi_spin_init(&ep->lpe_bplock);
-		/* create a buffer pool for the receive requests */
-		bp_attrs.size = sizeof(struct lnx_rx_entry);
-		bp_attrs.alignment = 8;
-		bp_attrs.max_cnt = UINT16_MAX;
-		bp_attrs.chunk_cnt = 64;
-		bp_attrs.flags = OFI_BUFPOOL_NO_TRACK;
-		rc = ofi_bufpool_create_attr(&bp_attrs, &ep->lpe_recv_bp);
-		if (rc) {
-			FI_WARN(&lnx_prov, FI_LOG_FABRIC,
-				"Failed to create receive buffer pool");
-			return -FI_ENOMEM;
-		}
+		dlist_init(&cep->cep_av_entry);
+		cep->cep_domain = cd;
+		cep->cep_parent = lep;
+
+		if (fclass == FI_CLASS_EP)
+			rc = fi_endpoint(cd->cd_domain, cd->cd_info,
+					 &cep->cep_ep, context);
+		else
+			rc = -FI_EINVAL;
+		if (rc)
+			goto fail;
```

Comment:
If you check for the fclass outside of the loop and exit before initializing any EPs, you can cleanup this error checking and remove the goto fail